Std/Data/List/Count.lean

```diff
@@ -0,0 +1,261 @@
+  rw [count, countp_eq_length]
+  refine ⟨fun h b hb => ?h₁, fun h b hb => ?h₂⟩
+  · refine' Eq.symm _
+    have := h b hb
+    simp at this
+    exact this
+  · rw [h b hb, beq_self_eq_true]
+
+@[simp] theorem count_replicate_self (a : α) (n : Nat) : count a (replicate n a) = n :=
+  (count_eq_length.2 <| fun _ h => (eq_of_mem_replicate h).symm).trans (length_replicate ..)
+
+theorem count_replicate (a b : α) (n : Nat) : count a (replicate n b) = if a = b then n else 0 := by
+  split
+  exacts [‹a = b› ▸ count_replicate_self .., count_eq_zero.2 <| mt eq_of_mem_replicate ‹a ≠ b›]
+
+theorem filter_beq' (l : List α) (a : α) : l.filter (· == a) = replicate (count a l) a := by
+  simp only [count, countp_eq_length_filter, eq_replicate, mem_filter, beq_iff_eq]
+  exact ⟨trivial, fun _ h => h.2⟩
+
+theorem filter_eq' (l : List α) (a : α) : l.filter (· = a) = replicate (count a l) a :=
+  filter_beq' l a
+
+theorem filter_eq (l : List α) (a : α) : l.filter (a = ·) = replicate (count a l) a := by
+  have := filter_eq' l a
+  simp only [eq_comm] at this ⊢
+  exact this
+
+theorem filter_beq (l : List α) (a : α) : l.filter (a == ·) = replicate (count a l) a :=
+  filter_eq l a
+
+theorem le_count_iff_replicate_sublist {l : List α} : n ≤ count a l ↔ replicate n a <+ l :=
+  ⟨fun h =>
+    ((replicate_sublist_replicate a).2 h).trans <| filter_eq l a ▸ filter_sublist _,
+    fun h => by
+      have := h.count_le a
+      simp only [count_replicate_self] at this ⊢
+      exact this⟩
+
+theorem replicate_count_eq_of_count_eq_length {l : List α} (h : count a l = length l) :
+    replicate (count a l) a = l :=
+  (le_count_iff_replicate_sublist.mp (Nat.le_refl _)).eq_of_length <|
+    (length_replicate (count a l) a).trans h
+
+@[simp] theorem count_filter {l : List α} (h : p a) : count a (filter p l) = count a l := by
+  rw [count, countp_filter]
+  congr
+  funext b
+  rw [(by rfl : (b == a) = decide (b = a))]
+  rw [decide_eq_decide]
+  simp; intro heq; exact heq ▸ h
+
+theorem count_le_count_map [DecidableEq β] (l : List α) (f : α → β) (x : α) :
+    count x l ≤ count (f x) (map f l) := by
+  rw [count, count, countp_map]
+  exact countp_mono_left <| by simp (config := {contextual := true})
+
+theorem count_erase (a b : α) :
+    ∀ l : List α, count a (l.erase b) = count a l - if a = b then 1 else 0
+  | [] => by simp
+  | c :: l => by
+    rw [erase_cons]
+    by_cases hc : c = b
+    · rw [if_pos hc, hc, count_cons, Nat.add_sub_cancel]
+    · rw [if_neg hc, count_cons, count_cons, count_erase a b l]
+      by_cases ha : a = b
+      · rw [← ha, eq_comm] at hc
+        rw [if_pos ha, if_neg hc, Nat.add_zero, Nat.add_zero]
+      · rw [if_neg ha, Nat.sub_zero, Nat.sub_zero]
+
+@[simp] theorem count_erase_self (a : α) (l : List α) : count a (List.erase l a) = count a l - 1 := by
+  rw [count_erase, if_pos rfl]
+
+@[simp] theorem count_erase_of_ne (ab : a ≠ b) (l : List α) : count a (l.erase b) = count a l :=
+  by rw [count_erase, if_neg ab, Nat.sub_zero]
```

Comment:
Same as `count_cons`, I feel like we could mark `count_erase` simp and delete these.